File: cockpit/api/cockpit_api/schemas/templates.py
"""Schemata für die globalen Vorlagen (ADR-0018)."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from cockpit_api.models.template import TemplateAudience
from cockpit_api.models.tenant import TenantProfile


class PlatformTemplateOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    key: str
    language: str
    subject: str | None
    body_html: str
    may_override: bool
    version: int
    audience: TemplateAudience
    audience_profile: TenantProfile | None
    is_active: bool
    #: Bei `audience = selection` die gewählten Kunden, sonst leer. Als eigenes
    #: Feld und nicht in `audience`: eine Oberfläche, die die Auswahl anzeigen
    #: will, soll dafür nicht eine zweite Anfrage brauchen.
    tenant_ids: list[UUID] = []
    updated_at: datetime
    updated_by: str | None


class PlatformTemplateSave(BaseModel):
    subject: str | None = Field(default=None, max_length=512)
    body_html: str
    may_override: bool = True
    audience: TemplateAudience = TemplateAudience.all
    audience_profile: TenantProfile | None = None
    #: `null` heisst „Auswahl nicht anfassen"; `[]` heisst „Auswahl leeren".
    #: Der Unterschied ist derselbe wie bei `clear_rbac` in den Einstellungen —
    #: ohne ihn gäbe es keinen Weg zurück.
    tenant_ids: list[UUID] | None = None
    is_active: bool = True
    actor: str = Field(min_length=1, max_length=320)


class PlatformTemplateListOut(BaseModel):
    items: list[PlatformTemplateOut]
    #: Die Schlüssel und Sprachen, die es gibt — damit die Oberfläche die
    #: Auswahl nicht doppelt pflegt.
    keys: list[str]
    languages: list[str]


__all__ = ["PlatformTemplateListOut", "PlatformTemplateOut", "PlatformTemplateSave"]

File: cockpit/web/src/api/templates.ts
import { del, get, put } from "./client";
import type { TenantProfile } from "./tenants";

/** An wen sich eine Vorlage richtet (ADR-0018 D5). */
export type TemplateAudience = "all" | "profile" | "selection";

export interface PlatformTemplate {
  key: string;
  language: string;
  subject: string | null;
  body_html: string;
  /** `false` = die Fassung gilt beim Kunden auch dann, wenn er eine eigene hat. */
  may_override: boolean;
  /** Steigt nur bei einer inhaltlichen Änderung (ADR-0018 D4). */
  version: number;
  audience: TemplateAudience;
  audience_profile: TenantProfile | null;
  is_active: boolean;
  /** Bei `audience = "selection"` die gewählten Kunden, sonst leer. */
  tenant_ids: string[];
  updated_at: string;
  updated_by: string | null;
}

export interface PlatformTemplateList {
  items: PlatformTemplate[];
  /** Die Schlüssel und Sprachen kommen von der API, damit die Liste nicht doppelt gepflegt wird. */
  keys: string[];
  languages: string[];
}

export interface PlatformTemplateSave {
  subject: string | null;
  body_html: string;
  may_override: boolean;
  audience: TemplateAudience;
  audience_profile: TenantProfile | null;
  /** `null` = Auswahl nicht anfassen, `[]` = Auswahl leeren. */
  tenant_ids: string[] | null;
  is_active: boolean;
  actor: string;
}

export function listPlatformTemplates(): Promise<PlatformTemplateList> {
  return get("/api/platform/templates");
}

export function savePlatformTemplate(
  key: string,
  language: string,
  body: PlatformTemplateSave,
): Promise<PlatformTemplate> {
  return put(`/api/platform/templates/${key}/${language}`, body);
}

export function deletePlatformTemplate(key: string, language: string): Promise<void> {
  return del(`/api/platform/templates/${key}/${language}`);
}
